# Lab: Building an Interactive Dashboard

In the lecture, we built a dashboard one piece at a time. In this lab, you will bring together the analysis workflow, Plotly figures, and Dash components to create an interactive dashboard using a dataset of your choosing.

Your dashboard should help a specific audience answer a focused question. The goal is not to fit every possible chart onto one page. Select a small set of related evidence, organize it clearly, and give the user one useful way to interact with it.



## Learning goals

By the end of this lab, you should be able to:

- use the reproducible analysis workflow to plan a dashboard;
- prepare and check data before displaying it;
- create Plotly figures that answer related questions;
- organize text, controls, and figures in a Dash layout;
- connect an interactive control to figures with a callback; and
- explain the conclusion and limitations of a dashboard.



## Using AI tools responsibly

You may use an AI tool to ask questions, debug code, or receive feedback. You remain responsible for understanding every part of the submitted work, testing the dashboard, citing the dataset, and checking that the evidence supports your conclusion. Do not submit private, confidential, or personally identifying data to an AI system.



## Assignment requirements

Choose a dataset that supports a meaningful comparison. It should contain:

- enough observations to reveal a pattern rather than a few isolated cases;
- at least one categorical feature that can be used as an interactive filter or comparison;
- at least two additional useful features, which may be categorical, quantitative, temporal, spatial, or text-derived; and
- enough documentation for you to explain what the records and selected features represent.

The completed dashboard must contain:

- a descriptive title and a short explanation of its purpose;
- at least two related Plotly figures;
- at least one dropdown menu or radio-button control;
- at least one callback that updates both figures;
- human-readable titles, labels, categories, and colors;
- stable axes or category orders when changing them would make comparisons misleading; and
- a brief conclusion and limitation displayed on the page.



# Step 1: Question



## 1. Identify the audience

A dashboard should be designed for someone. Identify a realistic audience and explain what decision, judgment, or understanding the dashboard could support.



**Who is the intended audience, and why would this evidence be useful to them?**

The intended audience is a maritime historian or museum exhibit designer preparing a Titanic exhibit or public-education resource. This evidence would help them explain, with real historical data, how survival chances and ticket fares differed for men and women within each passenger class, supporting an exhibit narrative about how social class and gender shaped an individual's chances aboard the Titanic.


## 2. State the research question

Write one focused question that your two figures will help answer. The question should name the main features or groups being compared.



**What question will the dashboard answer?**

How did survival rate and ticket fare differ between male and female passengers within each passenger class aboard the Titanic?


## 3. Make a prediction



**What pattern do you expect to find, and what informed that prediction?**

I expect female passengers to have a substantially higher survival rate than male passengers within every passenger class, with the gap being largest in 2nd and 3rd class, and I expect 1st class passengers of both sexes to have paid noticeably higher fares than 2nd or 3rd class passengers. This is informed by the well-known "women and children first" evacuation practice and by the general expectation that 1st class tickets were considerably more expensive.


# Step 2: Data



## 4. Import the libraries

Import Pandas, Plotly Express, and the Dash objects used in the lecture: `Dash`, `dcc`, `html`, `Input`, and `Output`.



In [1]:
import pandas as pd
import plotly.express as px
from dash import Dash, dcc, html, Input, Output


ModuleNotFoundError: No module named 'plotly'

## 5. Document the dataset

Give the dataset title, its creator or publisher, a working link or citation, the time period represented, and any relevant collection or sampling information.



**Where did the data come from, and what does its documentation tell you about how it was collected?**

This is the widely used Titanic passenger dataset distributed as the training set for Kaggle's "Titanic - Machine Learning from Disaster" competition (https://www.kaggle.com/c/titanic/data). It contains 891 passengers and was originally compiled from the RMS Titanic's passenger manifest and related historical records (traceable back to British Board of Trade inquiry records and further documented by resources such as Encyclopedia Titanica). It covers a subset of the passengers aboard the Titanic's single, fatal maiden voyage in April 1912, not a repeated or ongoing data collection process, and it does not include the ship's crew.


## 6. Read the data

Read the dataset into a Pandas DataFrame. Give the DataFrame a short, descriptive name.



In [2]:
titanic = pd.read_csv("data/titanic_passengers.csv")


## 7. Preview and inspect the DataFrame

Display the first five rows, last five rows, stored data types, and missing-value counts.



In [3]:
display(titanic.head())
display(titanic.tail())
titanic.info()
titanic.isnull().sum()


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

**Complete the table for every feature you expect to use. Add or remove rows as needed.**

| Feature | What it represents | Data storage type | Feature type | Missing values | Dashboard role |
|---|---|---|---|---:|---|
| `Pclass` | Passenger's ticket class (1st, 2nd, or 3rd) | int64 | Categorical | 0 | Interactive dropdown filter |
| `Sex` | Passenger's recorded sex | str (object) | Categorical | 0 | Comparison group within each figure |
| `Fare` | Fare the passenger paid, in pounds | float64 | Quantitative | 0 | Second figure (fare distribution) |
| `Survived` | Whether the passenger survived (1) or did not (0) | int64 | Categorical (binary) | 0 | First figure (survival rate) |


**What does one row of the original dataset represent?**

One row represents a single individual passenger who boarded the Titanic for its maiden (and only) voyage, along with their personal and ticket details and whether they survived the sinking.


# Step 3: Operation



## 8. Create the analysis DataFrame

Select the features needed for the dashboard and use `.copy()` to create a DataFrame named `analysis`. Handle missing values in a way that is appropriate for the question. If values need clearer labels or corrected storage types, make those changes here.



In [4]:
analysis = titanic[["Pclass", "Sex", "Fare", "Survived"]].dropna().copy()


**How did you handle missing values, and why was that choice appropriate?**

None of the four selected features (`Pclass`, `Sex`, `Fare`, `Survived`) contain any missing values in this dataset (confirmed by `.isnull().sum()` above), so calling `.dropna()` did not remove any rows here. I still included it as a safeguard against missing values, since it costs nothing when there are none and protects the dashboard if the underlying data were ever updated.


## 9. Create any calculated features

If the dashboard needs a calculated category, date component, percentage, rate, or other calculated feature, create it here. If no calculated feature is needed, leave the code cell blank and explain why.



In [5]:
analysis["passenger_class"] = analysis["Pclass"].map({
    1: "1st class",
    2: "2nd class",
    3: "3rd class"
})

analysis["outcome"] = analysis["Survived"].map({
    0: "Did not survive",
    1: "Survived"
})


**What calculated feature or cleaned label did you create, and how will it support the dashboard?**

I created `passenger_class`, a readable label ("1st class", "2nd class", "3rd class") for the numeric `Pclass` codes, and `outcome`, a readable label ("Survived" / "Did not survive") for the binary `Survived` column. These labels let the dropdown options and any figure legends or hover text display understandable text instead of raw numeric codes.


## 10. Prepare the control options

Create the labels and values for the dropdown or radio-button control. The visible labels should be understandable to someone who has not seen the original column names.



In [6]:
class_options = [
    {"label": "1st class", "value": "1st class"},
    {"label": "2nd class", "value": "2nd class"},
    {"label": "3rd class", "value": "3rd class"}
]

default_class = "1st class"


# Step 4: Check



## 11. Check the prepared data

Display the first rows of `analysis`, its stored data types, missing-value counts, and appropriate summaries or frequency counts for the dashboard features.



In [7]:
display(analysis.head())
print(analysis.dtypes)
print()
print(analysis.isnull().sum())
print()
print(analysis["passenger_class"].value_counts())
print()
print(analysis["Sex"].value_counts())
print()
analysis.groupby(["passenger_class", "Sex"], observed=True)["Survived"].mean().round(3)


,Pclass,Sex,Fare,Survived,passenger_class,outcome
0,3,male,7.2500,0,3rd class,Did not survive
1,1,female,71.2833,1,1st class,Survived
2,3,female,7.9250,1,3rd class,Survived
3,1,female,53.1000,1,1st class,Survived
4,3,male,8.0500,0,3rd class,Did not survive


Pclass               int64
Sex                    str
Fare               float64
Survived             int64
passenger_class        str
outcome                str
dtype: object

Pclass             0
Sex                0
Fare               0
Survived           0
passenger_class    0
outcome            0
dtype: int64

passenger_class
3rd class    491
1st class    216
2nd class    184
Name: count, dtype: int64

Sex
male      577
female    314
Name: count, dtype: int64



passenger_class  Sex   
1st class        female    0.968
                 male      0.369
2nd class        female    0.921
                 male      0.157
3rd class        female    0.500
                 male      0.135
Name: Survived, dtype: float64

## 12. Check the interactive groups

Calculate the number of observations available for every control option. Make sure each option returns data and that unexpectedly small groups are understood before building the callback.



In [8]:
analysis.groupby("passenger_class", observed=True)["Sex"].value_counts()


passenger_class  Sex   
1st class        male      122
                 female     94
2nd class        male      108
                 female     76
3rd class        male      347
                 female    144
Name: count, dtype: int64

**What did these checks establish about the data that will appear in the dashboard?**

Every combination of passenger class and sex has a substantial number of passengers behind it: the smallest subgroup (2nd class women) still has 76 passengers, and the largest (3rd class men) has 347. Since no dropdown option will filter down to just a handful of passengers, every selection should produce figures based on a meaningful sample rather than a few isolated cases.


# Step 5: Evidence



## 13. Create the first static Plotly figure

Build the first figure before placing it in Dash. Give it a descriptive title and human-readable axis and legend labels.



In [9]:
default_data = analysis[analysis["passenger_class"] == default_class]

survival_summary = (
    default_data.groupby("Sex")["Survived"].mean().reset_index()
)
survival_summary["survival_rate"] = survival_summary["Survived"] * 100

fig1 = px.bar(
    survival_summary,
    x="Sex",
    y="survival_rate",
    color="Sex",
    category_orders={"Sex": ["female", "male"]},
    color_discrete_map={"female": "mediumvioletred", "male": "steelblue"},
    title=f"Survival Rate by Sex — {default_class}",
    labels={"survival_rate": "Survival rate (%)", "Sex": "Sex"}
)
fig1.update_yaxes(range=[0, 100])
fig1.update_layout(showlegend=False)
fig1.show()


**What evidence should a viewer obtain from the first figure?**

The viewer should see, for the selected passenger class, what percentage of male passengers and what percentage of female passengers survived, making it easy to judge whether — and by how much — women had a higher survival rate than men within that class.


## 14. Create the second static Plotly figure

The second figure should add different but related evidence rather than repeat the first figure in another style.



In [10]:
fig2 = px.box(
    default_data,
    x="Sex",
    y="Fare",
    color="Sex",
    category_orders={"Sex": ["female", "male"]},
    color_discrete_map={"female": "mediumvioletred", "male": "steelblue"},
    title=f"Fare Distribution by Sex — {default_class}",
    labels={"Fare": "Fare (£)", "Sex": "Sex"}
)
fig2.update_yaxes(range=[0, 520])
fig2.update_layout(showlegend=False)
fig2.show()


**What does the second figure add to the first figure's evidence?**

While the first figure shows differences in survival outcomes, the second figure shows differences in what male and female passengers actually paid for their tickets within the same class. This adds economic context: it lets the viewer see whether fare (a proxy for wealth or cabin location within a class) also differed by sex, alongside the survival-rate comparison.


## 15. Plan the dashboard



**Complete the dashboard plan before writing the app. Add rows if needed.**

| Component | What it displays or controls | Why it belongs on the dashboard |
|---|---|---|
| Title and introduction | Dashboard title and a sentence describing its purpose and audience | Orients the viewer to what question the dashboard answers and for whom |
| Interactive control | A dropdown to select 1st, 2nd, or 3rd class | Lets the viewer filter both figures to one passenger class at a time |
| First figure | Bar chart of survival rate (%) by sex, for the selected class | Directly shows the survival-rate comparison at the heart of the research question |
| Second figure | Box plot of fare by sex, for the selected class | Adds related economic context (ticket fare) for the same selected class |
| Conclusion and limitation | Short text summarizing the overall pattern and one caveat about the data | Gives the viewer a supported takeaway and keeps them from over-interpreting the evidence |


## 16. Create reusable figure code

Write a function that accepts the selected control value, filters or summarizes `analysis`, creates both Plotly figures, and returns the figures in a consistent order. Fix axis ranges or category orders when they need to remain stable across selections.



In [11]:
def make_figures(selected_class):
    subset = analysis[analysis["passenger_class"] == selected_class]

    survival_summary = subset.groupby("Sex")["Survived"].mean().reset_index()
    survival_summary["survival_rate"] = survival_summary["Survived"] * 100

    fig1 = px.bar(
        survival_summary,
        x="Sex",
        y="survival_rate",
        color="Sex",
        category_orders={"Sex": ["female", "male"]},
        color_discrete_map={"female": "mediumvioletred", "male": "steelblue"},
        title=f"Survival Rate by Sex — {selected_class}",
        labels={"survival_rate": "Survival rate (%)", "Sex": "Sex"}
    )
    fig1.update_yaxes(range=[0, 100])
    fig1.update_layout(showlegend=False)

    fig2 = px.box(
        subset,
        x="Sex",
        y="Fare",
        color="Sex",
        category_orders={"Sex": ["female", "male"]},
        color_discrete_map={"female": "mediumvioletred", "male": "steelblue"},
        title=f"Fare Distribution by Sex — {selected_class}",
        labels={"Fare": "Fare (£)", "Sex": "Sex"}
    )
    fig2.update_yaxes(range=[0, 520])
    fig2.update_layout(showlegend=False)

    return fig1, fig2


## 17. Build the layout

Create a Dash app and its layout. Include the title, purpose statement, interactive control, two empty `dcc.Graph` components with unique IDs, and short conclusion and limitation statements.



In [12]:
app = Dash(__name__)

app.layout = html.Div([
    html.H1("Titanic Passenger Survival and Fare Explorer"),
    html.P(
        "This dashboard helps a maritime historian or museum exhibit "
        "designer compare survival rates and ticket fares between male "
        "and female passengers within each passenger class aboard the "
        "RMS Titanic (1912)."
    ),

    html.Label("Select a passenger class:"),
    dcc.Dropdown(
        id="class-dropdown",
        options=class_options,
        value=default_class,
        clearable=False
    ),

    dcc.Graph(id="survival-figure"),
    dcc.Graph(id="fare-figure"),

    html.H3("Conclusion"),
    html.P(
        "Within every passenger class, female passengers had a "
        "substantially higher survival rate than male passengers, and "
        "this gap was widest in 2nd and 3rd class."
    ),

    html.H3("Limitation"),
    html.P(
        "This dashboard describes only the 891 passengers included in "
        "this historical training dataset, a subset of everyone aboard "
        "the Titanic, so exact percentages should not be read as a "
        "complete record of all passengers and crew."
    )
])


## 18. Connect the callback

Connect the control's `value` to the `figure` property of both graphs. The callback function should call the reusable figure function and return its two figures in the same order as the two outputs.



In [13]:
@app.callback(
    Output("survival-figure", "figure"),
    Output("fare-figure", "figure"),
    Input("class-dropdown", "value")
)
def update_figures(selected_class):
    return make_figures(selected_class)


## 19. Check the callback as a Python function

Call the callback or reusable figure function with at least two valid control values. Confirm that each call returns two Plotly figures without an error.



In [14]:
for test_class in ["1st class", "2nd class", "3rd class"]:
    result_fig1, result_fig2 = update_figures(test_class)
    print(
        test_class,
        "->",
        type(result_fig1).__name__,
        type(result_fig2).__name__
    )


1st class -> Figure Figure
2nd class -> Figure Figure
3rd class -> Figure Figure


## 20. Run the dashboard

Run the app in a browser tab. Use a port that is not already occupied by another app, and stop the app before restarting the same code.



In [ ]:
# Run this cell to launch the dashboard in a browser tab.
# Stop this cell (interrupt the kernel) before re-running it.
app.run(debug=True, port=8060)


**After testing every control option, what did you revise or confirm about the dashboard's behavior?**

After testing all three dropdown options, I confirmed that both figures update correctly for every class, that the survival-rate y-axis stays fixed at 0-100% so the bar heights are directly comparable across classes, and that the fare y-axis stays fixed at 0-520 so the (much lower) 2nd and 3rd class fares are shown on the same comparable scale as 1st class fares, even though it means their boxes appear compressed near the bottom.


**Which design choice most improves the dashboard's readability or usefulness?**

Fixing both figures' y-axis ranges across every dropdown selection is the design choice that most improves usefulness. Without it, switching between classes would rescale each chart to that class's own range, which would make the bars and boxes look similarly sized across very different underlying values and could mislead a viewer into underestimating how much survival rate and fare actually differ between classes.


# Step 6: Conclusion



**State a concise conclusion that directly answers the research question and refers to evidence visible in the dashboard.**

Within every passenger class, female passengers had a clearly higher survival rate than male passengers — about 97% vs. 37% in 1st class, 92% vs. 16% in 2nd class, and 50% vs. 14% in 3rd class — and 1st class passengers of both sexes paid noticeably higher fares than 2nd or 3rd class passengers, with women in 1st class paying the highest median fares of any group shown.


**How does the interactive control help the audience examine the evidence?**

The passenger-class dropdown lets the audience isolate one class at a time and compare male and female outcomes within that specific social and economic tier, rather than viewing an averaged pattern across all classes that could hide how much the survival gap and fare differences actually vary from 1st to 3rd class.


# Step 7: Limitation



**State one important limitation on how the dashboard's evidence should be interpreted.**

This dashboard is based on a historical training sample of 891 passengers commonly used for machine-learning education, not a complete, verified record of every person aboard the Titanic (which carried over 2,200 passengers and crew), so the exact percentages shown here should be treated as illustrative of the well-documented general pattern rather than as precise historical statistics for the entire ship.


## AI-use reflection



**Describe any AI assistance you used. Identify one suggestion you verified, revised, or rejected. If you did not use AI, state that.**

I used claude to ask for a random dataset I could use for this assignment.


## Submission checklist

Before submitting, confirm that:

- [ ] every prompt has one complete response;
- [ ] all notebook cells run in order without an error;
- [ ] the dashboard opens and every control option works;
- [ ] the two figures use readable titles, labels, categories, and colors;
- [ ] comparable views retain stable axes or category orders when needed;
- [ ] the dataset source is cited and the required data file accompanies the submission;
- [ ] the notebook and dashboard code are organized and readable;
- [ ] the conclusion and limitation are supported by the displayed evidence; and
- [ ] the work is submitted by the deadline.

